File: packages/flutter-client/lib/src/debug/sandbox_user_record.dart
import 'package:collection/collection.dart';

String? _asString(dynamic v) => v == null ? null : (v is String ? v : v.toString());

Map<String, dynamic>? _asStringMap(dynamic v) {
  if (v is! Map) return null;
  final result = <String, dynamic>{};
  for (final entry in v.entries) {
    if (entry.key != null) {
      result[entry.key.toString()] = entry.value;
    }
  }
  return result;
}

/// Represents a sandbox user record returned from the `auth.listUsers` RPC.
class SandboxUserRecord {
  final String uid;
  final String? email;
  final String? displayName;
  final String? photoURL;
  final String? tenantId;
  final Map<String, dynamic> customClaims;

  const SandboxUserRecord({
    required this.uid,
    this.email,
    this.displayName,
    this.photoURL,
    this.tenantId,
    this.customClaims = const {},
  });

  /// Parses a sandbox user record from bridge JSON wire format.
  factory SandboxUserRecord.fromMap(Map<dynamic, dynamic> map) {
    final uid = _asString(map['uid']) ?? '';
    final email = _asString(map['email']);
    final displayName = _asString(map['displayName']);
    final photoURL = _asString(map['photoURL']);
    final tenantId = _asString(map['tenantId']) ?? _asString(map['tenant']);
    final claims = _asStringMap(map['customClaims']) ??
        _asStringMap(map['claims']) ??
        <String, dynamic>{};

    return SandboxUserRecord(
      uid: uid,
      email: email,
      displayName: displayName,
      photoURL: photoURL,
      tenantId: tenantId,
      customClaims: claims,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is SandboxUserRecord &&
        other.uid == uid &&
        other.email == email &&
        other.displayName == displayName &&
        other.tenantId == tenantId &&
        const MapEquality().equals(other.customClaims, customClaims);
  }

  @override
  int get hashCode => Object.hash(
        uid,
        email,
        displayName,
        tenantId,
        const MapEquality().hash(customClaims),
      );
}

File: packages/swift-client/Sources/PyricDebugUI/SandboxUserRecord.swift
import Foundation
import PyricFirestore

/// Represents an available sandbox user returned from the `auth.listUsers` RPC.
public struct SandboxUserRecord: Identifiable, Sendable, Hashable {
    public let uid: String
    public let email: String?
    public let displayName: String?
    public let photoURL: String?
    public let tenantId: String?
    public let customClaims: [String: AnySendable]

    public var id: String { uid }

    public init(
        uid: String,
        email: String? = nil,
        displayName: String? = nil,
        photoURL: String? = nil,
        tenantId: String? = nil,
        customClaims: [String: AnySendable] = [:]
    ) {
        self.uid = uid
        self.email = email
        self.displayName = displayName
        self.photoURL = photoURL
        self.tenantId = tenantId
        self.customClaims = customClaims
    }

    /// Parses a sandbox user record from bridge JSON wire format.
    public static func fromWire(_ wire: AnySendable) -> SandboxUserRecord? {
        guard let dict = wire.dictionaryValue,
              let uid = dict["uid"]?.stringValue, !uid.isEmpty else {
            return nil
        }
        let email = dict["email"]?.stringValue
        let displayName = dict["displayName"]?.stringValue
        let photoURL = dict["photoURL"]?.stringValue
        let tenantId = dict["tenantId"]?.stringValue ?? dict["tenant"]?.stringValue
        let claims = dict["customClaims"]?.dictionaryValue ?? dict["claims"]?.dictionaryValue ?? [:]

        return SandboxUserRecord(
            uid: uid,
            email: email,
            displayName: displayName,
            photoURL: photoURL,
            tenantId: tenantId,
            customClaims: claims
        )
    }

    public func hash(into hasher: inout Hasher) {
        hasher.combine(uid)
    }

    public static func == (lhs: SandboxUserRecord, rhs: SandboxUserRecord) -> Bool {
        lhs.uid == rhs.uid
    }
}
